File: node/src/ClientSideCache.ts
/**
 * Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
 */

import { randomUUID } from "crypto";

import { EvictionPolicy } from "./EvictionPolicy.js";

/**
 * Configuration options for creating a ClientSideCache.
 */
export interface ClientSideCacheConfig {
    /**
     * Maximum memory limit for the cache in kilobytes.
     * Must be greater than zero.
     */
    maxCacheKb: number;

    /**
     * Time-To-Live for cache entries in milliseconds.
     * Set to 0 to disable TTL expiration (entries remain until evicted or invalidated).
     */
    entryTtlMs: number;

    /**
     * Optional eviction policy to use when cache reaches memory limit.
     * Defaults to LRU if not specified.
     */
    evictionPolicy?: EvictionPolicy;

    /**
     * Whether to enable metrics collection for this cache.
     * Defaults to false if not specified.
     */
    enableMetrics?: boolean;
}

/**
 * Optional configuration options for the static create method.
 */
export interface ClientSideCacheOptions {
    /**
     * Optional eviction policy to use when cache reaches memory limit.
     */
    evictionPolicy?: EvictionPolicy;

    /**
     * Whether to enable metrics collection for this cache.
     */
    enableMetrics?: boolean;
}

/**
 * Configuration class for client-side caching.
 *
 * Client-side caching reduces network round-trips and server load by storing
 * frequently accessed data locally on the client. This class provides
 * configurable TTL-based expiration, multiple eviction policies, and
 * comprehensive metrics tracking.
 *
 * @example
 * ```typescript
 * // Create cache with auto-generated ID
 * const cache = ClientSideCache.create(1024, 60000); // 1MB cache, 1 min TTL
 *
 * // Create cache with custom configuration
 * const customCache = new ClientSideCache({
 *   maxCacheKb: 2048,
 *   entryTtlMs: 300000,
 *   evictionPolicy: EvictionPolicy.LFU,
 *   enableMetrics: true
 * });
 * ```
 */
export class ClientSideCache {
    /**
     * @internal
     * Unique identifier for the cache instance. Auto-generated using UUID.
     */
    readonly cacheId: string;

    /**
     * Maximum memory limit for the cache in kilobytes.
     */
    readonly maxCacheKb: number;

    /**
     * Time-To-Live for cache entries in milliseconds. 0 means no expiration.
     */
    readonly entryTtlMs: number;

    /**
     * Optional eviction policy to use when cache reaches memory limit.
     */
    readonly evictionPolicy?: EvictionPolicy;

    /**
     * Whether metrics collection is enabled for this cache.
     */
    readonly enableMetrics: boolean;

    /**
     * Creates a new ClientSideCache instance.
     *
     * @param config - Configuration options for the cache
     * @throws {Error} If maxCacheKb is not a positive number
     * @throws {Error} If entryTtlMs is negative
     */
    constructor(config: ClientSideCacheConfig) {
        if (config.maxCacheKb <= 0) {
            throw new Error("maxCacheKb must be a positive number");
        }

        if (config.entryTtlMs < 0) {
            throw new Error(
                "entryTtlMs must be non-negative (0 = no expiration)",
            );
        }

        this.cacheId = randomUUID();
        this.maxCacheKb = config.maxCacheKb;
        this.entryTtlMs = config.entryTtlMs;
        this.evictionPolicy = config.evictionPolicy;
        this.enableMetrics = config.enableMetrics ?? false;
    }

    /**
     * Factory method to create a ClientSideCache with auto-generated cache ID.
     *
     * @param maxCacheKb - Maximum memory limit for the cache in kilobytes
     * @param entryTtlMs - TTL for cache entries in milliseconds. Use 0 for no expiration.
     * @param options - Optional configuration options
     * @returns A new ClientSideCache instance with auto-generated cache ID
     * @throws {Error} If maxCacheKb is not a positive number
     * @throws {Error} If entryTtlMs is negative
     *
     * @example
     * ```typescript
     * // Simple cache with 1MB limit and no TTL
     * const cache = ClientSideCache.create(1024, 0);
     *
     * // Cache with TTL and LFU eviction
     * const cacheWithOptions = ClientSideCache.create(2048, 300000, {
     *   evictionPolicy: EvictionPolicy.LFU,
     *   enableMetrics: false
     * });
     * ```
     */
    static create(
        maxCacheKb: number,
        entryTtlMs: number,
        options?: ClientSideCacheOptions,
    ): ClientSideCache {
        return new ClientSideCache({
            maxCacheKb,
            entryTtlMs,
            evictionPolicy: options?.evictionPolicy,
            enableMetrics: options?.enableMetrics,
        });
    }
}

File: go/config/ClientSideCache.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package config

import (
	"fmt"

	"github.com/google/uuid"
	"github.com/valkey-io/valkey-glide/go/v2/internal/protobuf"
)

// EvictionPolicy represents the eviction policy for client-side cache entries.
// When the cache reaches its maximum size, it must evict existing entries to make room for new ones.
type EvictionPolicy int

const (
	// EvictionPolicyLRU (Least Recently Used) evicts the least recently accessed entry.
	// Best for recency-biased workloads like event streams and job queues.
	EvictionPolicyLRU EvictionPolicy = 0
	// EvictionPolicyLFU (Least Frequently Used) evicts the least frequently accessed entry.
	// Best for frequency-biased workloads like user profiles and product catalogs.
	EvictionPolicyLFU EvictionPolicy = 1
)

// String returns the string representation of the EvictionPolicy.
func (e EvictionPolicy) String() string {
	switch e {
	case EvictionPolicyLRU:
		return "LRU"
	case EvictionPolicyLFU:
		return "LFU"
	default:
		return fmt.Sprintf("EvictionPolicy(%d)", int(e))
	}
}

// ClientSideCache represents configuration for client-side caching with TTL-based expiration.
//
// This struct configures a local cache that stores read command responses
// on the client side to reduce network round-trips and server load. The cache
// uses Time-To-Live (TTL) based expiration, where entries are automatically
// removed after a specified duration.
//
// Important:
//   - Glide currently supports TTL-based caching only. Invalidation-based
//     client-side caching (where the server notifies clients of key changes) is not
//     currently supported. This means cached values may become stale if updated on
//     the server before the TTL expires.
//   - Currently, Glide's client-side cache supports lazy eviction only. Expired entries
//     are removed only when accessed after their TTL has expired. There is no proactive
//     background cleanup of expired entries.
//   - Currently, only read commands that retrieve entire values are cached (GET, HGETALL, SMEMBERS).
type ClientSideCache struct {
	// cacheId is a unique identifier for the cache instance, used internally to
	// identify the cache in the Rust core. Multiple clients sharing the same
	// ClientSideCache instance will use the same cacheId and therefore share
	// cached data and metrics.
	cacheId string

	// MaxCacheKb is the maximum size of the cache in kilobytes (KB).
	// This limits the total memory used by cached keys and values.
	// When this limit is reached, entries are evicted based on the eviction policy.
	MaxCacheKb uint64

	// EntryTtlMs is the Time-To-Live for cached entries in milliseconds.
	// After this duration, entries automatically expire and are removed from the cache.
	// Set to 0 to disable TTL expiration (entries remain until evicted or invalidated).
	EntryTtlMs uint64

	// EvictionPolicy is the policy for evicting entries when the cache reaches its maximum size.
	// If nil, the default policy of LRU will be used.
	EvictionPolicy *EvictionPolicy

	// EnableMetrics enables collection of cache metrics such as hit/miss rates.
	EnableMetrics bool
}

// NewClientSideCache creates a new ClientSideCache configuration with an auto-generated unique ID.
//
// This function configures a local cache that stores read command responses
// on the client side to reduce network round-trips and server load. The cache
// uses Time-To-Live (TTL) based expiration, where entries are automatically
// removed after a specified duration.
//
// Important:
//   - Glide currently supports TTL-based caching only. Invalidation-based
//     client-side caching (where the server notifies clients of key changes) is not
//     currently supported. This means cached values may become stale if updated on
//     the server before the TTL expires.
//   - Currently, Glide's client-side cache supports lazy eviction only. Expired entries
//     are removed only when accessed after their TTL has expired. There is no proactive
//     background cleanup of expired entries.
//   - Currently, only read commands that retrieve entire values are cached (GET, HGETALL, SMEMBERS).
//
// Note: In order for 2 clients to share the same cache, they must be
// created with the same ClientSideCache instance.
//   - Clients with different ClientSideCache instances will have separate caches,
//     even if the configurations are identical.
//   - Clients using different db's cannot share the same cache.
//   - Clients using different ACL users cannot share the same cache.
//
// Parameters:
//   - maxCacheKb: Maximum size of the cache in kilobytes (KB). Must be positive.
//   - entryTtlMs: Time-To-Live for cached entries in milliseconds. Set to 0 to disable TTL expiration.
//
// Returns:
//   - *ClientSideCache: A new ClientSideCache instance with auto-generated cache ID.
//   - error: An error if maxCacheKb is 0.
//
// Example:
//
//	cache, err := NewClientSideCache(1024, 60000) // 1 MB cache, 1 minute TTL
//	if err != nil {
//	    log.Fatal(err)
//	}
//	cache.WithEvictionPolicy(EvictionPolicyLRU).WithMetrics(true)
func NewClientSideCache(maxCacheKb uint64, entryTtlMs uint64) (*ClientSideCache, error) {
	if maxCacheKb == 0 {
		return nil, fmt.Errorf("maxCacheKb must be positive")
	}

	return &ClientSideCache{
		cacheId:        uuid.New().String(),
		MaxCacheKb:     maxCacheKb,
		EntryTtlMs:     entryTtlMs,
		EvictionPolicy: nil,
		EnableMetrics:  false,
	}, nil
}

// WithEvictionPolicy sets the eviction policy for the cache.
// Returns the same ClientSideCache instance for method chaining.
func (c *ClientSideCache) WithEvictionPolicy(policy EvictionPolicy) *ClientSideCache {
	c.EvictionPolicy = &policy
	return c
}

// WithMetrics enables or disables cache metrics collection.
// Returns the same ClientSideCache instance for method chaining.
func (c *ClientSideCache) WithMetrics(enable bool) *ClientSideCache {
	c.EnableMetrics = enable
	return c
}

// toProtobuf converts the ClientSideCache configuration to protobuf format.
// This method is used internally to serialize the cache configuration for
// communication with the Rust core.
func (c *ClientSideCache) toProtobuf() *protobuf.ClientSideCache {
	protoCache := &protobuf.ClientSideCache{
		CacheId:       c.cacheId,
		MaxCacheKb:    c.MaxCacheKb,
		EntryTtlMs:    c.EntryTtlMs,
		EnableMetrics: c.EnableMetrics,
	}

	if c.EvictionPolicy != nil {
		switch *c.EvictionPolicy {
		case EvictionPolicyLRU:
			policy := protobuf.EvictionPolicy_LRU
			protoCache.EvictionPolicy = &policy
		case EvictionPolicyLFU:
			policy := protobuf.EvictionPolicy_LFU
			protoCache.EvictionPolicy = &policy
		}
	}

	return protoCache
}
